### ISSUE TITLE
support for v3 auth is part broken, part incomplete

### ISSUE THREAD

**Issue by majewsky** [gophercloud/gophercloud#252]
Our corporate OpenStack installations support Identity v3 exclusively, and they use the whole feature set, down to users from one domain having roles in projects in different domains. When I look at [`func AuthOptionsFromEnv`](https://github.com/gophercloud/gophercloud/blob/50cdddf51c54e8df178af065055cbc5b3cf442e1/openstack/auth_env.go#L15), about half of the environment variables that I need are missing, and for some, the implementation in Gophercloud is wrong.

I'll recap the terminology of the OpenStack Identity API here in case some readers are not familiar with it: In the v2 API, users are limited to *tenants*, and users can only do work in their respective tenant, so the sufficient auth options are:

```go
UserName   string // from $OS_USERNAME
UserID     string // from $OS_USERID
TenantName string // from $OS_TENANT_NAME
TenantID   string // from $OS_TENANT_ID
```

That's also what's in `type AuthOptions` (plus endpoint URL plus password). In the v3 API, users reside in *domains*, and tenants (now called *projects*) do so, too. A token can be scoped to a domain, to a project in a domain (not necessarily the user's domain), or to nothing. The sufficient auth options in this case are: [(reference)](https://github.com/openstack/python-openstackclient/blob/83cf09df28e200ba40f403504db3b11e623978fd/doc/source/man/openstack.rst#environment-variables)

```go
UserName          string // from $OS_USERNAME
UserID            string // from $OS_USERID
UserDomainName    string // from $OS_USER_DOMAIN_NAME
UserDomainID      string // from $OS_USER_DOMAIN_ID
//for domain scope:
DomainName        string // from $OS_DOMAIN_NAME
DomainID          string // from $OS_DOMAIN_ID
//for project scope:
ProjectName       string // from $OS_PROJECT_NAME
ProjectID         string // from $OS_PROJECT_ID
ProjectDomainName string // from $OS_PROJECT_DOMAIN_NAME
ProjectDomainID   string // from $OS_PROJECT_DOMAIN_ID
```

Note how there is nearly no overlap with the v2 variable names. This is not reflected in `func AuthOptionsFromEnv` at all. It uses `OS_TENANT_{ID,NAME}` for the project scope, even though this is not consistent with the upstream `openstack` CLI tool. And even worse, it uses `OS_DOMAIN_{ID,NAME}` to identify the domain of the user (and *also* the domain to scope to, if no project is given). This is completely wrong, and breaks cross-domain scoping usecases, i.e. obtaining a token for one domain or a project within that domain, as a user who resides in a different domain.

## Possible solutions

The main shortcoming is that [`type AuthOptions`](https://github.com/gophercloud/gophercloud/blob/50cdddf51c54e8df178af065055cbc5b3cf442e1/auth_options.go#L13) lacks a number of fields to express all the authentication scenarios that we need, and the fields `.Domain{Name,ID}` are confusingly named because they reflect `OS_USER_DOMAIN_{NAME,ID}` instead of `OS_DOMAIN_{NAME,ID}`.

### Option 1a

If backwards compatibility were not a concern (I actually don't know if it is for you), here's what I would do:
* Rename `.Domain{Name,ID}` to `.UserDomain{Name,ID}` as a clarification.
* Move `.Tenant{Name,ID}` into a `type V2 struct`, so they would be referenced as `opts.V2.Tenant{Name,ID}`. These fields would thus be documented as being relevant for v2 auth only.
* Add a `type V3 struct` containing `.{Domain,Project,ProjectDomain}{Name,ID}` fields that are only considered for v3 auth.

That way, if there is ever a v4 auth, we would not need to go through the same mess again. `type AuthOptions` would just gain a new field of type `type V4 struct` that has all the parameters that v4 auth cares about.

### Option 1b

Alternatively, replace the `type AuthOptions` by multiple types `AuthOptionsV2`, `AuthOptionsV3`, and so on, which all implement a shared interface (which would then be what `func AuthOptionsFromEnv` returns, for instance). This solution is not entirely backwards-compatible, since all the functions that take or return `type AuthOptions` would need to be adjusted to take/return the interface instead. Although, without actually having done it, it could probably be done so that at least a sizable number of consumers of this module would still compile. If `type AuthOptions` is left intact for backwards compatibility, deprecation warnings should be issued if it is used.

### Option 2

If backwards compatibility is a major concern, here's an alternative proposal:
* Reuse `.Tenant{Name,ID}` for the `OS_PROJECT_{NAME,ID}` in v3 auth. It will be confusing in the long term because the "tenant" term is being replaced by "project" across the OpenStack documentation and source code, but v2 tenants and v3 projects are technically equivalent, so it's still appropriate.
* Add `.TenantDomain{Name,ID}` to mirror `OS_PROJECT_DOMAIN_{NAME,ID}`. (It should be called `.ProjectDomain{Name,ID}`, but if we're already misspelling "project" as "tenant", let's at least be consistent about it.)
* Add `.ScopeDomain{Name,ID}` to mirror `OS_DOMAIN_{NAME,ID}`. The name `.Domain{Name,ID}` would be more appropriate, but it's already taken by `OS_USER_DOMAIN_{NAME,ID}`.

You see how ugly this gets... So, what's your choice? 😁 

Note: I'm willing to do a pull request to try to rectify things, once a solution is chosen.

**Comment by jrperritt** [comment]
Let me think about it.

@jtopjian Any thoughts on this? Have you run into it yet?

**Comment by jtopjian** [comment]
I haven't run into any v3 issues, no. I've never been in a position to use cross-domain scoping, but short of that, all other auth methods have worked for me.

A few months ago, someone brought up v3 issues with Terraform. I spent a few hours comparing how v3 is implemented in Gophercloud to what the OpenStack/Keystone docs and code say. The scoping (sans cross-domain) took me for a loop, but in the end it looked correct. The only "issue" I saw was the reuse of the word `tenant` instead of `project`, but I figured that was rather minor.

I asked the reporter to try again, and all was reported working with no changes made.

It sounds like there are two topics here:

* The way v2 and v3 are structured in Gophercloud seems a bit tangled and hard to follow. This might cause future maintainability problems. For example, when I was reviewing the code, I was confused as to why it seemed like the stuff in `auth_options.go` is duplicating what is found in the respective `identity` packages, but upon closer review, I saw the differences and it ended up making sense.
* There might be missing support for cross-domain scoping, but maybe it's possible, but not immediately obvious due to the above.

Maybe a good way to approach this is to break down each way of authenticating to Keystone with Gophercloud (token, user/pass, user/pass/scope, cross-domain, etc) with example code. It might bring about any missing functionality, as well as serve as reference documentation.

**Comment by majewsky** [comment]
I would like to get back to this, but my plate is currently more than full. For the record, in the application that I'm now working on, I've worked around the missing cross-domain scoping by implementing my own `identity/v3/tokens.AuthOptionsBuilder`: https://github.com/sapcc/limes/blob/44d6fc8d897d4d5eb231983d6360bbad7111ff5b/pkg/limes/config.go#L239-L266

**Comment by majewsky** [comment]
Yes, that should fix that particular problem with that method.

FYI, here's what [Limes](https://github.com/sapcc/limes) (and AFAIK also [Hermes](https://github.com/sapcc/hermes)) is currently using: https://github.com/sapcc/limes/blob/master/pkg/limes/auth.go

**Comment by jtopjian** [comment]
> Yes, that should fix that particular problem with that method.

@majewsky which problems would remain if that was implemented? The tenant/project naming is certainly one. But specifically with scoping - anything else?

@jobrs Thanks for the link. Would it be beneficial to also add that to the other [`auth_options.go`](https://github.com/gophercloud/gophercloud/blob/master/auth_options.go#L318)?

If you're able to add some unit tests for this, definitely open an official PR. If you don't have time or aren't really interested, I can take over.

As well, could either of you provide some basic openstack commands to create a simple reproducible scenario? Something from the very beginning would be best (creation of domains, projects, and users, and what the auth/env settings look like when working in the different scopes). It'd be really appreciated since my current understanding of cross-domain scoping is really limited.

**Comment by jobrs** [comment]
@jtopjian Yes, the error case should be entirely removed (i.e. remove the message and all references)

Is there an existing unit-test that could be adjusted?

As scenario for domain-scoping is the creation of new OpenStack projects (tenants). For example here: http://www.florentflament.com/blog/setting-keystone-v3-domains.html --> section "Creating projects and users"

**Comment by jobrs** [comment]
@jtopjian pull-request provided. The test errors are found also in the master branch, so they are independent of the proposed change.

@majewsky do you see more issues?

**Comment by jtopjian** [comment]
@jobrs Thank you for submitting the PR. I totally apologize for dropping the ball on this - it kept getting pushed further down my inbox :(

I will review this in detail as soon as I get a chance.

**Comment by jtopjian** [comment]
To provide an update to this issue: #489 was merged which allows `scope['domain']['name']` to be set.

Unless I am mistaken, Gophercloud now supports all possible auth and scope options. The way those options are set, though, may not be immediately recognizable to some and differs from the "standard" implementation. 

In order to fully consider this issue resolved, the following needs investigated/completed:

1. Have separate `TenantID`, `TenantName`, and `ProjectID`, `ProjectName`.
2. Add support for the `OS_PROJECT_DOMAIN_*` environment variables.

I can add these items to my to-do, but unless someone reports hard breakage or incompatibility, it'll have a low-priority, since all auth/scope combinations should now currently be possible. I would be happy to review a PR if someone wanted to look at doing this in the meantime.

**Comment by jobrs** [comment]
:+1:

**Comment by Adrian-Turjak** [comment]
@jtopjian So to continue our discussion I guess the thing we want to consider is how gophercloud should handle rescoping of a token.

So from what I gather, gophercloud has scoping for password auth, but not token auth. So a lot of the same mechanisms could be reused for that?

Eventually down the line you'll also want to add support for more than just password and token auth methods, especially as the other MFA stuff starts coming in and being used. Hopefully by then we'll get a spec written that helps people writing client libraries...

**Comment by jtopjian** [comment]
Actually before we get to that, I think we need to make some changes to `AuthOptions` in order to handle scoping correctly.

I might be jumping ahead of myself, but per discussion in https://github.com/terraform-providers/terraform-provider-openstack/issues/275, it looks as though `gophercloud.AuthOptions` needs to support the ability to pass in a Domain Name or ID specifically for scoping purposes.

Per our _previous_ discussion, Gophercloud does indeed support scoping (https://github.com/gophercloud/gophercloud/blob/master/auth_options.go#L265), but it tries to determine the scope from the options given in `gophercloud.AuthOptions`. This struct only accepts _one_ set of `DomainName` or `DomainID`. Therefore, passing a second set for scoping purposes isn't possible right now.

Again, I hope I'm not jumping ahead of myself, but this is where the inability to do cross-domain scoping comes in. As wonky as Gophercloud looks to people familiar with the way the Python SDK works, specifying what would be `OS_USER_DOMAIN_*` as `DomainID` or `DomainName` actually works out for all scenarios _but_ cross-domain scoping.

So, I think that somewhere in [`AuthOptions`](https://github.com/gophercloud/gophercloud/blob/master/auth_options.go#L30) we need to provide the ability to specify a second set of Domain ID/Name (let's call it `ScopedDomainID` and `ScopedDomainName`). If these are set, they take precedence during the scoping logic. But if they aren't set, the existing scope determination logic will still work for backwards compatibility.

I really hope I'm not off mark here. If so, then I'm _really_ lost.

**Comment by majewsky** [comment]
That sounds about right. The `AuthOptions` have two different concerns, identifying the user and identifying the scope. In general, both are not connected in any way. Only by convention, user and project are *usually* in the same domain, which is why the existing `AuthOptions` works in many cases.

For perspective, if I were to design `AuthOptions` from scratch today, it would probably be an interface, not a struct. And it would possibly be a pair of two interfaces to separate the user-identification and scope-identification concerns. Although "password" and "token" are the most common authentication methods, Keystone can also optionally support SAML, OAuth, even SSL client certs, etc. Stuffing all of that into one struct (plus all the various options for scope) is bound to get messy. Though I guess you won't have much of a choice here if you want to maintain at least some backwards-compatibility.

**Comment by jtopjian** [comment]
> I were to design AuthOptions from scratch today, it would probably be an interface, not a struct

Actually, there already is an interface:

https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/tokens/requests.go
https://github.com/gophercloud/gophercloud/blob/781450b3c4fcb4f5182bcc5133adb4b2e4a09d1d/openstack/client.go#L192

So if an application can accurately detect if Identity v3 is being used (config flag or something), the app can bypass `gophercloud.AuthOptions` altogether. This way, the application can correctly build a scope rather than have `gophercloud.AuthOptions` try to do it.

The unfortunate part of this is that it makes `gophercloud.AuthOptions` moot and its existence confusing. I'm on the fence on whether to amend `gophercloud.AuthOptions` to do clever things or consider it deprecated for v3.

**Comment by majewsky** [comment]
If you consider `AuthOptions` deprecated, then what about `AuthOptionsFromEnv`? This function is extremely valuable for *if* it can cover all the environment variables that the upstream openstackclient recognizes (which it does not, at the moment). Maybe a better API would be:

```go
func AuthenticateFromEnv(client *gophercloud.ProviderClient) error
func AuthenticatedClientFromEnv() (*gophercloud.ProviderClient, error)
```

Compared to `AuthOptionsFromEnv`, this could recognize whether v2 or v3 auth needs to be used, and use either `AuthenticateV2` or `AuthenticateV3` internally with an appropriate implementation of `tokens3.AuthOptionsBuilder` in the v3 case.

This API also has the benefit that, if and when Identity v4 appears, clients only have to update to a newer Gophercloud and get v4 auth without any changes in their own code.

**Comment by jtopjian** [comment]
IMO, `AuthOptionsFromEnv` should be moved to `gophercloud/utils`. Reading in environment variables is arguably more on the application side than a core part of Gophercloud. I think the function served a useful purpose being in Gophercloud "core", but as things grow, it makes more sense it move it somewhere more appropriate.

**Comment by jtopjian** [comment]
Hi all,

There have been two major updates to this:

The first is that #896 has been merged which adds support for a Scope struct to `gophercloud.AuthOptions`. As mentioned, the `v3/tokens` package already supported this but now that the `gophercloud.AuthOptions` struct supports this, is should be trivial to create all required scoping configurations.

Second, I have done a [major rewrite](https://github.com/gophercloud/utils/pull/27) of the `gophercloud/utils` `clientconfig` package. Initially, `clientconfig` was created to handle `clouds.yaml` support but I've extended it to act as a single point of entry that user-facing apps can use to authenticate with Gophercloud. It's in `utils` because it's meant as an app-facing helper and does not necessarily involve "core" API interaction with OpenStack. 

This does not make `gophercloud.AuthOptions` obsolete. `clientconfig` ultimately ends up using `gophercloud.AuthOptions` but it does it in a more friendly way by checking for explicitly defined credentials, `clouds.yaml` entries, and environment variables. In a way, it can be thought of as a successor to Gophercloud's `AuthOptionsFromEnv`.

I've done my best to make `clientconfig` conform to the correct rules of auth options and precedence with OpenStack. If there is a mistake, please let me know.

On the Terraform side, I will be implementing `clientconfig` as the standard way of authentication shortly.

I'm going to leave this issue open for a little longer for comments and for anyone to point out mistakes. At the moment, though, I consider all authentication issues to be resolved. Ideas and comments about making Gophercloud more flexible and extendible for future and custom authorizations are absolutely valid, but those discussions might be better suited in a new issue.

I appreciate everyone's input and, even more, patience, with this.
